src/protocol/boolean/bitwise_lt.rs

```diff
@@ -0,0 +1,258 @@
+use super::prefix_or::PrefixOr;
+use super::BitOpStep;
+use crate::error::BoxError;
+use crate::ff::BinaryField;
+use crate::protocol::{context::ProtocolContext, mul::SecureMul, RecordId};
+use crate::secret_sharing::Replicated;
+use futures::future::try_join_all;
+use std::iter::{repeat, zip};
+
+/// This is an implementation of Bitwise Less-Than on bitwise-shared numbers.
+///
+/// `BitwiseLessThan` takes inputs `[a]_B = ([a_1]_p,...,[a_l]_p)` where
+/// `a1,...,a_l ∈ {0,1} ⊆ F_p` and `[b]_B = ([b_1]_p,...,[b_l]_p)` where
+/// `b1,...,b_l ∈ {0,1} ⊆ F_p`, then computes `h ∈ {0, 1} <- a <? b` where
+/// `h = 1` iff `a` is less than `b`.
+///
+/// Note that `[a]_B` can be converted to `[a]_p` by `Σ (2^i * a_i), i=0..l`. In
+/// other words, if comparing two integers, the protocol expects inputs to be in
+/// the little-endian; the least-significant byte at the smallest address (0'th
+/// element).
+///
+/// 5.3 Bitwise Less-Than
+/// "Unconditionally Secure Constant-Rounds Multi-party Computation for Equality, Comparison, Bits, and Exponentiation"
+/// I. Damgård et al.
+pub struct BitwiseLessThan<'a, B: BinaryField> {
+    a: &'a [Replicated<B>],
+    b: &'a [Replicated<B>],
+}
+
+impl<'a, B: BinaryField> BitwiseLessThan<'a, B> {
+    #[allow(dead_code)]
+    pub fn new(a: &'a [Replicated<B>], b: &'a [Replicated<B>]) -> Self {
+        debug_assert_eq!(a.len(), b.len(), "Length of the input bits must be equal");
+        Self { a, b }
+    }
+
+    /// Step 1. `for i=0..l-1, [e_i] = XOR([a_i], [b_i])`
+    ///
+    /// We can compute `[a] ⊕ [b], a,b ∈ Z_2` as `[a + b]` locally.
+    ///
+    /// # Example
+    /// ```ignore
+    ///   //  bit-0         bit-7
+    ///   //    v             v
+    ///   [a] = 1 0 1 0 1 0 0 0   // 21 in little-endian
+    ///   [b] = 0 1 1 1 1 0 0 0   // 30 in little-endian
+    ///   [e] = 1 1 0 1 0 0 0 0
+    /// ```
+    fn step1(a: &[Replicated<B>], b: &[Replicated<B>]) -> Vec<Replicated<B>> {
+        zip(a, b)
+            .map(|(&a_bit, &b_bit)| a_bit + b_bit)
```

Comment:
I think this will fail in any field other than F2. The paper suggests computing XOR as follows:

```
[d_i]_p = [a_i]_p - [b_i]_p
[e_i]_p = MULT([d_i]_p, [d_i]_p)
```

This logic is currently missing.

Reply:
Don't we already have an XOR protocol?  The modulus reduction does: $[d_i]_p = [a_i]_p + [b_i]_p - 2.MULT([a_i]_p,[b_i]_p)$.